internal/testhelpers/e2e/spawn.go

```diff
@@ -45,3 +51,29 @@ func BackgroundProcess() SpawnOptions {
 		return nil
 	}
 }
+
+type Shell string
+
+const (
+	Bash Shell = "bash"
+	Zsh        = "zsh"
+	Tcsh       = "tcsh"
+	Fish       = "fish"
+	Cmd        = "cmd.exe"
+)
+
+func WithShell(shell Shell, s *Session) SpawnOptions {
+	return func(opts *Options) error {
+		if len(opts.Options.Args) == 0 {
+			return errs.New("e2e.WithShell must come after e2e.WithArgs")
+		}
+		opts.Options.CmdName = string(shell)
+		shellArg := "-c"
+		if shell == Cmd {
+			shellArg = "/k"
+		}
+		cmd := s.Exe + " " + strings.Join(opts.Options.Args, " ")
+		opts.Options.Args = []string{shellArg, cmd} // e.g. -c "state activate project/org"
```

Comment:
```suggestion
		opts.Options.Args = append([]string{shellArg, opts.Options.CmdName}, opts.Options.Args...) // e.g. -c "state activate project/org"
```

Note I also replaced `s.Exe` with `opts.Options.CmdName`; s.Exe is not always used for all spawn commands, but CmdName should be far as I can tell.

Reply:
That won't work because `opts.Options.CmdName` is the name of the shell we're running. Also, all arguments need to be grouped into a single quoted argument. Appending args into a `[]string` will unquote/ungroup them and you'll wind up with an incorrect command like `bash -c state activate project/org` instead of `bash -c "state activate project/org"`.

Reply:
I think just string joining and using quotes is gonna be very error prone.

My solution might not be appropriate, but I'd still like us to address these issues in this PR:

- Passing arguments via strings.Join and manually quoting it is going to create a lot of gotcha scenarios. Can we find a way to do this cleaner?
- On that point; `bash -c` has some issues, as raised during our stand. I think perhaps a more appropriate solution would be to use a custom SpawnWithCmd(bash). Then once spawned you can use ts.Send(opts.Options.CmdName), though we'd still be facing the arg parsing problem raised first.
- We can't just assume s.Exe is the command that's being spawned

Reply:
I've addressed the first point with the proper shell quoting of arguments. I've addressed point three by allowing the tester to specify an exe, but the `WithShell()` option silently invokes the shell with that exe.

As for option 2, I see two competing use cases:

1. Launch a single command in a particular shell and then exit.
2. Launch a shell, run 1 or more commands in it, and then exiting the shell (a lot like the `state activate` or `state shell` workflow).

I believe this ticket and this functionality should be targetted at (1). Therefore I am not in favor of having the test writer launch the shell, `Send()`ing input to it, and then `Send()`ing an exit command. It also feels like scope creep for this ticket. For more elaborate scenarios, maybe we should be leveraging different GitHub shells for running our integration tests instead of assuming bash for everything.

Reply:
The way I see it not addressing point 2 creates three issues:

- It makes for an error-prone API, because as a developer I can run `WithShell()` and start sending a bunch of input (ie. `tp.Send()`) which it won't respect implicitly because it only runs a single command.
- It means that we're not actually emulating what a user would do, which is the point of our integration tests. A user isn't going to run `bash -c whatever`, they will open a bash shell and then run `whatever`.
- The existing issue regarding loading RC files with `-c`

I don't think this is adding scope creep to the story, as the point of the story is to test that we actually can run the supported shells on the supported platforms, which the current iteration of the PR is not adequately doing.

Reply:
Fair enough. The existing `SpawnInShell()` function threw me off. I have removed it in favor of `SpawnShell()` which spawns an interactive shell that you can send commands to and read output from.